### PR TITLE
Worldpay: Adding support for google pay and apple pay

### PR DESCRIPTION
**Pull request by Heavyblade**
Summary:
------------------------------
This PR adds support for the apple pay payment method on the Worldpay
adapter by adding a new network token type

Test Execution:
------------------------------
Finished in 149.298941 seconds.
83 tests, 331 assertions, 2 failures, 0 errors, 0 pendings, 0 omissions, 0 notifications
97.5904% passed

RuboCop:
------------------------------
723 files inspected, no offenses detected

### CODE REVIEW

File: test/unit/gateways/worldpay_test.rb [hunk 1]
```diff
@@ -14,11 +14,11 @@ def setup
...
-      year: 2020,
-      first_name: 'John',
-      last_name: 'Smith',
-      verification_value: '737',
-      brand: 'elo')
+       year: 2020,
+       first_name: 'John',
+       last_name: 'Smith',
+       verification_value: '737',
+       brand: 'elo')
```
Comment (curiousepic) [inline test/unit/gateways/worldpay_test.rb:21]: Strange whitespace edit, was Rubocop triggering on this?

File: lib/active_merchant/billing/gateways/worldpay.rb [hunk 2]
```diff
@@ -813,6 +819,10 @@ def card_code_for(payment_method)
       def eligible_for_0_auth?(payment_method, options = {})
         payment_method.is_a?(CreditCard) && %w(visa master).include?(payment_method.brand) && options[:zero_dollar_auth]
       end
+
+      def card_holder_name(payment_method, options)
+        test? && options[:execute_threed] && !options[:three_ds_version]&.start_with?('2') ? '3D' : payment_method.name
+      end
```
Comment (curiousepic) [inline lib/active_merchant/billing/gateways/worldpay.rb:825]: Nice

File: lib/active_merchant/billing/gateways/worldpay.rb [hunk 3]
```diff
@@ -441,15 +441,23 @@ def add_payment_method(xml, amount, payment_method, options)
...
-          xml.tag! 'EMVCO_TOKEN-SSL', 'type' => 'NETWORKTOKEN' do
+          xml.tag! 'EMVCO_TOKEN-SSL', 'type' => token_type do
             xml.tokenNumber payment_method.number
             xml.expiryDate do
               xml.date(
                 'month' => format(payment_method.month, :two_digits),
                 'year' => format(payment_method.year, :four_digits_year)
               )
             end
+            card_h_name = card_holder_name(payment_method, options)
```
Comment (curiousepic) [inline lib/active_merchant/billing/gateways/worldpay.rb:459]: This mix of two words and one letter `card_h_name` is a little out of place; we usually avoid abbreviations. I assume you didn't want to mirror the method name, so in this case just using something like `name` would make sense.

File: lib/active_merchant/billing/gateways/worldpay.rb [hunk 4]
```diff
@@ -441,15 +441,23 @@ def add_payment_method(xml, amount, payment_method, options)
       end
 
       def add_network_tokenization_card(xml, payment_method)
+        token_type = {
+          apple_pay: 'APPLEPAY',
+          google_pay: 'GOOGLEPAY',
+          network_token: 'NETWORKTOKEN'
+        }[payment_method.source]
```
Comment (curiousepic) [inline lib/active_merchant/billing/gateways/worldpay.rb:448]: We usually add something like this as a Constant map, similar to `CARD_CODES` up above.

**Comment by Heavyblade** [comment]
@curiousepic thanks for your comments, they all have been addressed.

**Comment by curiousepic** [comment]
Awesome - merged!

**Pull request closed without merging**